src/module_library/root_onedim.h

```diff
@@ -0,0 +1,515 @@
+ * double f(double x) { return x*x - 3; };
+ *
+ * //Declare the solver
+ * root_algorithm::root_finder<root_algorithm::secant> solver;
+ *
+ * // Call the solve method to find a root.
+ * root_algorithm::result_t result = solver.solve(f, 1., 2.);
+ * @endcode
+ *
+ * The function can be a c++ function, lambda function, or function object.
+ * However, for algorithms which explicitly evaluate derivatives, you must
+ * pass a function object with the appropriate derivative method.
+ *
+ */
+template <typename Method>
+struct root_finder : public Method {
+
+    size_t max_iterations = 100;
+    double _abs_tol = std::numeric_limits<double>::epsilon();
+    double _rel_tol = 2 * std::numeric_limits<double>::epsilon();
+
+    root_finder() {}
+    root_finder(size_t max_iter, double abs_tol, double rel_tol) : max_iterations{max_iter}, _abs_tol{abs_tol}, _rel_tol{rel_tol} {}
+
+    using flagged_state = with_flag<typename Method::state>;
+
+    template <typename F, typename... Args>
+    result_t solve(F&& func, Args&&... args)
+    {
+        flagged_state s = Method::initialize(
+            std::forward<F>(func), std::forward<Args>(args)...);
+
+        for (size_t i = 0; i < (max_iterations + 1); ++i) {
+            if (is_valid(s)) {
+                s.flag = Method::check_convergence(s.state, _abs_tol, _rel_tol);
+            }
+
+            if (s.flag != Flag::valid) {
+                return make_result(s, i);
+            }
+
+            s = Method::iterate(std::forward<F>(func), s.state);
+        }
+
+        s.flag = Flag::max_iterations;
+        return make_result(s, max_iterations);
+    }
+
+    private:
+
+    inline result_t make_result(const flagged_state& s, size_t iteration)
+    {
+        return result_t{Method::root(s.state), Method::residual(s.state), iteration, s.flag};
+    }
+
+};
+
+// Methods
+// Multi-step methods
+
+struct two_step_method {
+
+    struct state {
+        graph_t last;
+        graph_t current;
+    };
+
+    template <typename F>
+    with_flag<state> initialize(F&& fun, double x0, double x1)
+    {
+        graph_t first{x0, fun(x0)};
+        graph_t second{x1, fun(x1)};
+        return state{first, second};
+    }
+
+
+    Flag check_convergence(const state& s,  double abs_tol, double rel_tol){
+
+
+        if (is_zero(s.current.y, abs_tol)) {
+            return Flag::residual_zero;
+        }
+
+        if (is_close(s.last.x, s.current.x, abs_tol, rel_tol)) {
+            return Flag::delta_root_zero;
+        }
+
+        return Flag::valid;
+    }
+
+    inline double root(const state& s)
+    {
+        return s.current.x;
+    }
+
+    inline double residual(const state& s)
+    {
+        return s.current.y;
+    }
+
+};
+
+struct secant : public two_step_method {
+
+
+    template <typename F>
+    with_flag<state> iterate(F&& fun, const state& s)
+    {
+        double secant_slope = (s.current.y - s.last.y) / (s.current.x - s.last.x);
+        if (secant_slope == 0.0) {
```

Comment:
Usually this would check for a very small value instead of exact equality with zero. There are a few other cases like this in the code as well.